### ISSUE TITLE
[Enhancement] [WebAuthenticator] [iOS] Enable WebAuthenticator to launch using an ephemeral (private) web session

### ISSUE THREAD

**Issue by dominicpease** [xamarin/Essentials#1528]
### Description

Across a series of iOS platform versions, several web authentication strategies have been introduced that differ in user experinece, especially around cookie handling and user privacy. For example, my understanding is that older `SFSafariVewController`-based authentication operated in an environment where cookies were isolated to the app itself, preventing SSO across apps; newer methods ((`ASWebAuthenticationSession`/`SFAuthenticationSession`) by default share cookies with the host system, which allows for SSO but can complicate multi-user scenarios and also necessitate an alert popup ("[app] wants to use [site]") every time a session is initiated.

It turns out that `ASWebAuthenticationSession` has the capability (as of iOS 13) to allow the developer to choose whether to prioritize SSO and the benefits of a shared context or a multi-user-friendly and alert-less UX by using `ASWebAuthenticationSession.PrefersEphemeralWebBrowserSession`. If true, cookies are isolated as if in a private browsing mode, which also suppresses the alert popup; if false, it uses the default behavior and allows SSO.

This proposal is to allow Xamarin developers to configure the desired behavior through a static configuration property.

### Suggested implementation

I propose adding a new static read-only property to the `WebAuthenticator` class on iOS called `Preferences` initialized with an object of new type `WebAuthenticatorPreferences`. For now I propose only a single option, `bool EphemeralSession` with a default of `false`. If set to true,  `ASWebAuthenticationSession.PrefersEphemeralWebBrowserSession` will be set to true if the implementation uses that implementation type.

Beyond that, I suggest we consider changing the existing code flow such that, if `EphemeralSession` is set to `true`, we immediately fall back to `SFSafariViewController` to honor the cookie handling of the developer to the extent we can. I am less sure of the impact of this, so I would consider this an optional extension of the idea.

Finally, we could consider adding similar features to other platforms, and then perhaps offering an overload of the `AuthenticateAsync` method to configure this on a call-by-call basis. I consider this low priority compared to the steps above. 

If approved, I believe I could work on this proposal and submit a PR.

**Comment by dominicpease** [comment]
Sorry, new to this-- reading the contribution guidelines, I thought I was supposed to be waiting for this to be approved before starting work on it. Should I just do the work and submit a PR without that?

**Comment by dfarrNTST** [comment]
Yes, please get this approved and put into the library.  It is causing a lot of issues with user experience at the moment.

**Comment by PBhuva** [comment]
@dominicpease are there any updates on release of this enhancement? We want to go live with this SDK.

**Comment by monkeypro** [comment]
After some testruns it seems as _Password Autofill_ is inactivated on iOS if **PrefersEphemeralWebBrowserSession** is set to **true**

**Comment by dominicpease** [comment]
This issue seems to be addressed by Xamarin Essentials 1.7. See [this page for more details](https://devblogs.microsoft.com/xamarin/xamarin-essentials-1-7-and-introducing-net-maui-essentials/). When we have a chance, we will test in our app and close this issue if applicable.

**Comment by Px7-941** [comment]
@dominicpease  Thanks for the tip, I'll give it a quick test tomorrow.

**Comment by monkeypro** [comment]
I still have an issues that the _Save password dialog_ do not show up if **PrefersEphemeralWebBrowserSession** is set to **true**
If I have a password save from before it will be suggested but I am not asked if if I want to save a new login

**Comment by dominicpease** [comment]
@monkeypro This is by design. The point of the ephemeral web browser session is that it is ephemeral-- it comes and goes without leaving any record of its existence. It also corresponds to, for example, Incognito mode in Chrome.

If this is confusing to you, the important thing to understand is why Apple pops up the warning we use `PrefersEphemeralWebBrowserSession` to avoid. They do it because they think there's a security issue in how browsers handle security when not in Incognito mode. They reason that if apps can pop up browser sessions that interact with normal cookies and so forth, it might be possible to get a user logged in with less-than-complete user awareness of what's going on (and there is a certain amount of truth to this). So they provided the workaround of allowing the use of a private browsing session instead. The consequence is that you get the safety of not having potential 'gray-zone' security issues, but also aspects of private browsing that you don't want.

Apple's solution is somewhat deficient in various ways, but there's no real fix for the 'bug' you're talking about given the nature of this approach. You just need to choose which option for `PrefersEphemeralWebBrowserSession` is the lesser of two evils until Apple takes a different path.

**Comment by monkeypro** [comment]
Thank you @dominicpease for a detailed explanation!
It was my lack of understanding the work **ephemeral** that made me think this was a bug!

**Comment by jamesmontemagno** [comment]
Implemented in 1.7 :)
